### PR TITLE
SPI chaining with TMCStepper simplified interface

### PR DESCRIPTION
**Pull request by Bob-the-Kuhn**
@teemuatlut has decided to use the stepper object creation call to init the TMC chain index rather than using a seperate call.

PR #15081 implemented the Marlin code for the TMC chain interface that I proposed.  teemuatlut has decided to use a simpler method.  This PR implements the Marlin side of the simpler interface.

As with PR #15081 , this PR is the companion to the TMC library [PR #66](https://github.com/teemuatlut/TMCStepper/pull/66).  

NOTE - TMC library [PR #66](https://github.com/teemuatlut/TMCStepper/pull/66) may not be the one that actually gets merged.  I'll update this PR if that happens.

CHANGES
1. tmc_util.h - add chain init versions of TMC create object calls
2. trinamic.cpp  - a) add chain init versions of TMC create object calls, b) make the use of the new calls conditional on if a daisy chain exists and c) remove the chain init section..

NOTE - 2b is only needed until the new TMC library is available.  Once that is available we can delete the old ones in both files.

This code has been tested on a mixed TMC2130/TMC5160 chain on a DUE RAMPS-FD V2.1 system.

---

TMC2660 devices will not support daisy chaining right away.  It uses a 20 bit command/response structure while the TMC2130 family uses a 40 bit structure.  Modifications to the soft and hard SPI libraries will probably be needed to accommodate the 4 bit transfers.

### CODE REVIEW

**Comment by teemuatlut** [comment]
Could the whole `TMC_USE_CHAIN` macro be removed?
All the `*_CHAIN_POS` by default are `0` which means the TMCStepper library won't use chaining for that driver. It doesn't hurt anything if we pass this `0` argument to the class constructor.
EDIT: Following the established Marlin configuration style, I'd actually prefer if the default value be `-1`.

**Comment by thinkyhead** [comment]
> I'd actually prefer if the default value be -1.

Would you then be checking for `*_CHAIN_POS <= 0` in the constructor?
We do use -1 for "`NO_PIN`" but 0 can mean "none" in the context of a 1-based index.

**Comment by thinkyhead** [comment]
- Rebased on top of a commit removing old instances of `TMC_USE_CHAIN`.
- Added a commit removing the `TMC_USE_CHAIN` conditional.

**Comment by teemuatlut** [comment]
> Would you then be checking for *_CHAIN_POS <= 0 in the constructor?
We do use -1 for "NO_PIN" but 0 can mean "none" in the context of a 1-based index.

Functionally they're the same but ´-1´ would visually be a better indication.

The simplified logic behind the functionality
```cpp
int8_t link_index = -1; // Given in constructor. Default = -1
int8_t chain_length = 0;
int8_t i = 1;

if (link_index > chain_length)
  chain_length = link_index;

/* Send 40 bits of valid data */

while (i < link_index) { // Nothing if link_index = -1
  // Send 40 dummy bits
  i++;
}

// CS high
// CS low

while (i < chain_length) { // Nothing if chain_length = 0
  // Send 40 dummy bits
  i++;
}

/* Read 40 bits into MCU */

```

**Comment by thinkyhead** [comment]
> -1 would visually be a better indication

- Except, in Marlin our first brain reflex to "`#define OPTION -1`" is "must be a pin number."
- Revising the documentation _**again**_ is bothersome for those who maintain it.

**Comment by thinkyhead** [comment]
```
Marlin/src/module/stepper/../../feature/tmc_util.h:113:62: error: no matching function
  for call to 'TMC2130Stepper::TMC2130Stepper(const uint16_t&, const float&,
  const uint16_t&, const uint16_t&, const uint16_t&, const uint8_t&)'
       TMC(CS, RS, pinMOSI, pinMISO, pinSCK,  axis_chain_index)
```
It looks like the `TMCStepper` code is not yet brought up-to-date in PlatformIO.

File: Marlin/src/module/stepper/trinamic.cpp [hunk 1]
```diff
@@ -44,9 +44,9 @@ enum StealthIndex : uint8_t { STEALTH_AXIS_XY, STEALTH_AXIS_Z, STEALTH_AXIS_E };
 //   AI = Axis Enum Index
 // SWHW = SW/SH UART selection
 #if ENABLED(TMC_USE_SW_SPI)
-  #define __TMC_SPI_DEFINE(IC, ST, L, AI) TMCMarlin<IC##Stepper, L, AI> stepper##ST(ST##_CS_PIN, ST##_RSENSE, TMC_SW_MOSI, TMC_SW_MISO, TMC_SW_SCK)
+  #define __TMC_SPI_DEFINE(IC, ST, L, AI) TMCMarlin<IC##Stepper, L, AI> stepper##ST(ST##_CS_PIN, ST##_RSENSE, TMC_SW_MOSI, TMC_SW_MISO, TMC_SW_SCK, ST##_CHAIN_POS)
 #else
-  #define __TMC_SPI_DEFINE(IC, ST, L, AI) TMCMarlin<IC##Stepper, L, AI> stepper##ST(ST##_CS_PIN, ST##_RSENSE)
+  #define __TMC_SPI_DEFINE(IC, ST, L, AI) TMCMarlin<IC##Stepper, L, AI> stepper##ST(ST##_CS_PIN, ST##_RSENSE, ST##_CHAIN_POS)
```
Comment (thinkyhead) [inline Marlin/src/module/stepper/trinamic.cpp:49]: Just waiting for TMCStepper to get updated in PlatformIO-land and this new signature will compile ok.

**Comment by sl1pkn07** [comment]
Hi

seems needs rebase

greetings

**Comment by sl1pkn07** [comment]
with this pr applied, i get printer halt when do M112

I wired the drivers like 





and with this config

```diff
diff --git a/Marlin/Configuration_adv.h b/Marlin/Configuration_adv.h
index 5959de4e6..ba9db7c0a 100644
--- a/Marlin/Configuration_adv.h
+++ b/Marlin/Configuration_adv.h
@@ -1732,7 +1732,7 @@
     #define X_CURRENT     750  // (mA) RMS current. Multiply by 1.414 for peak current.
     #define X_MICROSTEPS   16  // 0..256
     #define X_RSENSE     0.11
-    #define X_CHAIN_POS     -1  // <=0 : Not chained. 1 : MCU MOSI connected. 2 : Next in chain, ...
+    #define X_CHAIN_POS     3  // <=0 : Not chained. 1 : MCU MOSI connected. 2 : Next in chain, ...
   #endif
 
   #if AXIS_IS_TMC(X2)
@@ -1746,7 +1746,7 @@
     #define Y_CURRENT     800
     #define Y_MICROSTEPS   16
     #define Y_RSENSE     0.11
-    #define Y_CHAIN_POS     -1
+    #define Y_CHAIN_POS     2
   #endif
 
   #if AXIS_IS_TMC(Y2)
@@ -1760,7 +1760,7 @@
     #define Z_CURRENT     850
     #define Z_MICROSTEPS   16
     #define Z_RSENSE     0.11
-    #define Z_CHAIN_POS     -1
+    #define Z_CHAIN_POS     1
   #endif
 
   #if AXIS_IS_TMC(Z2)
@@ -1781,7 +1781,7 @@
     #define E0_CURRENT    500
     #define E0_MICROSTEPS  16
     #define E0_RSENSE    0.11
-    #define E0_CHAIN_POS    -1
+    #define E0_CHAIN_POS    4
   #endif
 
   #if AXIS_IS_TMC(E1)
@@ -1823,14 +1823,14 @@
    * Override default SPI pins for TMC2130, TMC2160, TMC2660, TMC5130 and TMC5160 drivers here.
    * The default pins can be found in your board's pins file.
    */
-  //#define X_CS_PIN          -1
-  //#define Y_CS_PIN          -1
-  //#define Z_CS_PIN          -1
+  #define X_CS_PIN            P1_15
+  #define Y_CS_PIN            P1_15
+  #define Z_CS_PIN            P1_15
   //#define X2_CS_PIN         -1
   //#define Y2_CS_PIN         -1
   //#define Z2_CS_PIN         -1
   //#define Z3_CS_PIN         -1
-  //#define E0_CS_PIN         -1
+  #define E0_CS_PIN           P1_15
   //#define E1_CS_PIN         -1
   //#define E2_CS_PIN         -1
   //#define E3_CS_PIN         -1
```

also i have test with different CS pin

Rearm+Ramps 1.4+tmc2130spi

greetings

**Comment by thinkyhead** [comment]
> seems needs rebase

Github tells us that so you don't have to.

**Comment by Bob-the-Kuhn** [comment]
Here's a SPI daisy chain diagram using TMC's nomenclature.


image

Here's the `Configuration_adv.h` excerpts for this example:
```cpp
  #if AXIS_IS_TMC(X)
    #define X_CURRENT     800  // (mA) RMS current. Multiply by 1.414 for peak current.
    #define X_MICROSTEPS   16  // 0..256
    #define X_RSENSE     0.11
    #define X_CHAIN_POS     1  // 0 - Not chained, 1 - MCU MOSI connected, 2 - next in chain, ...
  #endif

  #if AXIS_IS_TMC(X2)
    #define X2_CURRENT    800
    #define X2_MICROSTEPS  16
    #define X2_RSENSE    0.11
    #define X2_CHAIN_POS    4
  #endif

  #if AXIS_IS_TMC(Y)
    #define Y_CURRENT     800
    #define Y_MICROSTEPS   16
    #define Y_RSENSE     0.11
    #define Y_CHAIN_POS     3
  #endif

  #if AXIS_IS_TMC(E0)
    #define E0_CURRENT    800
    #define E0_MICROSTEPS  16
    #define E0_RSENSE    0.11
    #define E0_CHAIN_POS    2
  #endif


  /**
   * Override default SPI pins for TMC2130, TMC2160, TMC2660, TMC5130 and TMC5160 drivers here.
   * The default pins can be found in your board's pins file.
   */
  #define X_CS_PIN          P1_15
  #define Y_CS_PIN          P1_15
  //#define Z_CS_PIN          -1
  #define X2_CS_PIN         P1_15
  //#define Y2_CS_PIN         -1
  //#define Z2_CS_PIN         -1
  //#define Z3_CS_PIN         -1
  #define E0_CS_PIN         P1_15
```

**Comment by sl1pkn07** [comment]
sorry but what is the difference? except you not use Z and use double X

the scheme is the same (except my example is taked from wikipedia) but look the same like you

**Comment by Bob-the-Kuhn** [comment]
I've gotten the TMC side (SDO, SDI) connected incorrectly in the past.  I keep incorrectly thinking that SDO on the TMC should be connected to MOSI from the CPU.  I was wondering if you were making the same mistake.

**Comment by sl1pkn07** [comment]
Hi

My mcu MOSI is connected to the tmc SDI

Greetings

**Comment by thinkyhead** [comment]
> sorry but what is the difference?

_"using TMC's nomenclature."_

**Comment by Bob-the-Kuhn** [comment]
@thinkyhead - the updated TMC library is now available in PlatformIO.  The ID number has not changed.

Travis did not automatically pull in the updated library.  Please point Travis at the new library & re-run.

**Comment by thinkyhead** [comment]
I'm not sure if this is Marlin or TMCStepper…

```
Marlin/src/module/stepper/trinamic.cpp:49:125: error: no matching function for call to 'TMCMarlin<TMC2660Stepper, 'E', '0', (AxisEnum)3u>::TMCMarlin(int, double, int)'
   #define __TMC_SPI_DEFINE(IC, ST, L, AI) TMCMarlin<IC##Stepper, L, AI> stepper##ST(ST##_CS_PIN, ST##_RSENSE, ST##_CHAIN_POS)
                                                                                                                             ^
Marlin/src/module/stepper/trinamic.cpp:55:37: note: in expansion of macro '__TMC_SPI_DEFINE'
 #define _TMC_SPI_DEFINE(IC, ST, AI) __TMC_SPI_DEFINE(IC, ST, TMC_##ST##_LABEL, AI)
                                     ^
Marlin/src/module/stepper/trinamic.cpp:56:32: note: in expansion of macro '_TMC_SPI_DEFINE'
 #define TMC_SPI_DEFINE(ST, AI) _TMC_SPI_DEFINE(ST##_DRIVER_TYPE, ST, AI##_AXIS)
                                ^
Marlin/src/module/stepper/trinamic.cpp:65:32: note: in expansion of macro 'TMC_SPI_DEFINE'
   #define TMC_SPI_DEFINE_E(AI) TMC_SPI_DEFINE(E##AI, E)
                                ^
Marlin/src/module/stepper/trinamic.cpp:92:3: note: in expansion of macro 'TMC_SPI_DEFINE_E'
   TMC_SPI_DEFINE_E(0);
   ^
In file included from Marlin/src/module/stepper/trinamic.h:35:0,
                 from Marlin/src/module/stepper/trinamic.cpp:32:
Marlin/src/module/stepper/../../feature/tmc_util.h:279:5: note: candidate: TMCMarlin<TMC2660Stepper, AXIS_LETTER, DRIVER_ID, AXIS_ID>::TMCMarlin(uint16_t, float, uint16_t, uint16_t, uint16_t) [with char AXIS_LETTER = 'E'; char DRIVER_ID = '0'; AxisEnum AXIS_ID = (AxisEnum)3u; uint16_t = unsigned int]
     TMCMarlin(const uint16_t CS, const float RS, const uint16_t pinMOSI, const uint16_t pinMISO, const uint16_t pinSCK) :
     ^
Marlin/src/module/stepper/../../feature/tmc_util.h:279:5: note:   candidate expects 5 arguments, 3 provided
Marlin/src/module/stepper/../../feature/tmc_util.h:276:5: note: candidate: TMCMarlin<TMC2660Stepper, AXIS_LETTER, DRIVER_ID, AXIS_ID>::TMCMarlin(uint16_t, float) [with char AXIS_LETTER = 'E'; char DRIVER_ID = '0'; AxisEnum AXIS_ID = (AxisEnum)3u; uint16_t = unsigned int]
     TMCMarlin(const uint16_t cs_pin, const float RS) :
     ^
Marlin/src/module/stepper/../../feature/tmc_util.h:276:5: note:   candidate expects 2 arguments, 3 provided
Marlin/src/module/stepper/../../feature/tmc_util.h:274:7: note: candidate: constexpr TMCMarlin<TMC2660Stepper, 'E', '0', (AxisEnum)3u>::TMCMarlin(const TMCMarlin<TMC2660Stepper, 'E', '0', (AxisEnum)3u>&)
 class TMCMarlin<TMC2660Stepper, AXIS_LETTER, DRIVER_ID, AXIS_ID> : public TMC2660Stepper, public TMCStorage<AXIS_LETTER, DRIVER_ID> {
       ^
Marlin/src/module/stepper/../../feature/tmc_util.h:274:7: note:   candidate expects 1 argument, 3 provided
Marlin/src/module/stepper/../../feature/tmc_util.h:274:7: note: candidate: constexpr TMCMarlin<TMC2660Stepper, 'E', '0', (AxisEnum)3u>::TMCMarlin(TMCMarlin<TMC2660Stepper, 'E', '0', (AxisEnum)3u>&&)
Marlin/src/module/stepper/../../feature/tmc_util.h:274:7: note:   candidate expects 1 argument, 3 provided
*** [.pio/build/megaatmega2560/src/src/module/stepper/trinamic.cpp.o] Error 1
```

**Comment by thinkyhead** [comment]
Have the chaining changes been moved into `master` and tagged with a new version number?

**Comment by Bob-the-Kuhn** [comment]
I think you are correct about the version number.  PlatformIO says the latest is 0.5.0 released 5 days ago.  The previous one was 0.4.6

**Comment by Bob-the-Kuhn** [comment]
I was able to reproduce the Travis error.

The error message is because Travis is using the old TMC library.  Once it pulls in 0.5.0 all is fine.

I've confirmed that the mod to platformio.ini pulls in 0.5.0 even if the older TMC library is prsent.

**Comment by sl1pkn07** [comment]
M112....m112....

ok

M122

~~~
M122
		X	Y	Z	E
Enabled		false	false	false	false
Set current	750	800	850	500
RMS current	734	795	826	489
MAX current	1035	1121	1165	689
Run current	23/31	25/31	26/31	15/31
Hold current	11/31	12/31	13/31	7/31
CS actual	11/31	12/31	13/31	7/31
PWM scale	4	4	28	4
vsense		1=.18	1=.18	1=.18	1=.18
stealthChop	true	true	true	true
msteps		16	16	16	16
tstep		max	max	max	max
pwm
threshold	98	98	658	64
[mm/s]		100	100	3	30
OT prewarn	false	false	false	false
OT prewarn has
been triggered	false	false	false	false
off time	4	4	4	4
blank time	24	24	24	24
hysteresis
-end		2	2	2	2
-start		1	1	1	1
Stallguard thrs	19	8	0	0
DRVSTATUS	X	Y	Z	E
stallguard				*	
sg_result	0	0	0	0
fsactive					
stst		*	*	*	*
olb		*	*		
ola		*	*		
s2gb					
s2ga					
otpw					
ot					
Driver registers:
		X	0xE0:0B:00:00
		Y	0xE0:0C:00:00
		Z	0x81:0D:00:00
		E	0x80:07:00:00


Testing X connection... OK
Testing Y connection... OK
Testing Z connection... OK
Testing E connection... OK
~~~

only Z can move (E0 is not tested) from the LCD. `G0 X<value>`, `G0 Y<value>` do nothing. `G0 Z<value>` works ok

All axes can do G28, and do ok the endstopless (diag1), but when walk to the center of the bed for probe the BLtouch, the X do strange sound and only walk the half of size he need to travel

The config is untouched (posted avobe)

greetings

**Comment by sl1pkn07** [comment]
more test. 

if first do `G28`, do like posted avobe
(poweroff/poweron)
if do `G28 X`, i can move the X axis without problem
if do `G28 Y`, i can move the Y axis without problem
if do `G0 X100` (after `G28X`), the X axis is move
if do `G0 Y100` (after `G28Y`), the Y axis is move
(back to 0 with `G0 X0` and then `G0 Y0`)
if do `G0 X200 Y200`, only Y is move. X stop working. do `G0 X200` do nothing
(poweroff/poweron)
if do `G28XY`, both go to home, if do `G0 X200` the X is move, if do `G28 Y200` the Y axis is move, if do `G28 X0 Y0`, only Y is move. X stop working

seems when run 2 commands in the same Gcode, only work the last and the first stop work

when run `G28` (all axis), X and Y stop to work, and only work Z

**Comment by Bob-the-Kuhn** [comment]
@sl1pkn07 - you have TMC communications which is a big step.

Next things I'd try:
- Issue M121 command to disable hardware endstops
- Use M211 to make sure the software endstops are disabled.
- Increase current by 50%.
- Verify that the SPI chain sequence in Configuration_adv.h is correct.  Change the microstep settings in Configuration_adv.h one by one from16 to 32 and back and see if the distance moved is cut in half for the correct axis.

**Comment by sl1pkn07** [comment]
Hi again

>Verify that the SPI chain sequence in Configuration_adv.h is correct. Change the microstep settings in Configuration_adv.h one by one from16 to 32 and back and see if the distance moved is cut in half for the correct axis.

this work as spected, i've change, one by one, the msteps for each driver and send movement. works ok (move the half distance in each try per driver)

the rest of the points tomorrow

greetings

**Comment by sl1pkn07** [comment]
Hi

im not sure what i need doing with the first steps

- M121 seems do nothing (no messages in console)
- M211 seems works
- i'm not sure what i need do with increase the current. change the current seems works because if do M122, the current is change)

greetings

greetings

**Comment by thinkyhead** [comment]
Rebased on the latest `bugfix-2.0.x`.
```
git fetch origin
git checkout "TMC-chain-(simplified-interface)"
git reset --hard "origin/TMC-chain-(simplified-interface)"
```

**Comment by thinkyhead** [comment]
> I've confirmed that the mod to platformio.ini pulls in 0.5.0 even if the older TMC library is prsent.

Weird. I have pulled in 0.5.0 but when I run `mftest mega 13` to build the SCARA test I still get the same error as before.

`Marlin/src/module/stepper/trinamic.cpp:49:125: error: no matching function for call to 'TMCMarlin<TMC2660Stepper, 'E', '0', (AxisEnum)3u>::TMCMarlin(int, double, int)'
   #define __TMC_SPI_DEFINE(IC, ST, L, AI) TMCMarlin<IC##Stepper, L, AI> stepper##ST(ST##_CS_PIN, ST##_RSENSE, ST##_CHAIN_POS)`

Is this an issue where the TMC2660 is not up to date with the rest of `TMCMarlin`?

**Comment by thinkyhead** [comment]
Note that the above test combines TMC 2209, 2130, and 2660 in a single configuration. Is it possible that this kind of setup is currently unsupported?

**Comment by teemuatlut** [comment]
There is no `stepperX(X_CS_PIN, X_RSENSE, X_CHAIN_POS)` constructor for TMCMarlin<TMC2660Stepper, AXIS_LETTER, DRIVER_ID, AXIS_ID>

This could fix it:
```diff
-    TMCMarlin(const uint16_t cs_pin, const float RS) :
+    TMCMarlin(const uint16_t cs_pin, const float RS, const uint8_t) :
      TMC2660Stepper(cs_pin, RS)
      {}
-    TMCMarlin(const uint16_t CS, const float RS, const uint16_t pinMOSI, const uint16_t pinMISO, const uint16_t pinSCK) :
+    TMCMarlin(const uint16_t CS, const float RS, const uint16_t pinMOSI, const uint16_t pinMISO, const uint16_t pinSCK, const uint8_t) :
      TMC2660Stepper(CS, RS, pinMOSI, pinMISO, pinSCK)
      {}
```

SPI chaining isn't yet supported for TMC2660 and you will not be able to chain 2660 with other SPI drivers because the SPI calls are not the same size.

**Pull request merged**